src/probnum/filtsmooth/gaussfiltsmooth/kalman.py

```diff
@@ -24,406 +33,186 @@ class Kalman(BayesFiltSmooth):
+    predict
+        Custom prediction step. Choose between e.g. classical and square-root implementation. Default is 'predict_via_transition`.
+    measure
+        Custom measurement step. Choose between e.g. classical and square-root implementation. Default is 'measure_via_transition`.
+    update
+        Custom update step. Choose between e.g. classical, Joseph form, and square-root implementation. Default is 'update_classic`.
+    smooth_step
+        Custom update step. Choose between e.g. classical, and square-root implementation. Default is 'rts_smooth_step_classic`.
     """
 
-    def __init__(self, dynamics_model, measurement_model, initrv):
-        """Check that the initial distribution is Gaussian."""
-        if not issubclass(type(initrv), Normal):
-            raise ValueError(
-                "Gaussian filters/smoothers need initial "
-                "random variables with Normal distribution."
-            )
-        super().__init__(dynamics_model, measurement_model, initrv)
-
-    def filtsmooth(self, dataset, times, _intermediate_step=None):
-        """Apply Gaussian filtering and smoothing to a data set.
+    def __init__(
+        self,
+        dynamics_model,
+        measurement_model,
+        initrv,
+        predict=predict_via_transition,
+        measure=measure_via_transition,
+        update=update_classic,
+        smooth_step=rts_smooth_step_classic,
+    ):
+        self.dynamics_model = dynamics_model
+        self.measurement_model = measurement_model
+        self.initrv = initrv
+        self.predict = lambda *args, **kwargs: predict(dynamics_model, *args, **kwargs)
+        self.measure = lambda *args, **kwargs: measure(
+            measurement_model, *args, **kwargs
+        )
+        self.update = lambda *args, **kwargs: update(measurement_model, *args, **kwargs)
+        self.smooth_step = smooth_step
+        super().__init__(
+            dynamics_model=dynamics_model,
+            measurement_model=measurement_model,
+            initrv=initrv,
+        )
 
-        Parameters
-        ----------
-        dataset : array_like, shape (N, M)
-            Data set that is filtered.
-        times : array_like, shape (N,)
-            Temporal locations of the data points.
-            The zeroth element in times and dataset is the location of the initial random variable.
+    def iterated_filtsmooth(
+        self, dataset, times, stopcrit=None, _intermediate_step=None
+    ):
+        """Compute an iterated smoothing estimate with repeated posterior linearisation.
 
-        Returns
-        -------
-        KalmanPosterior
-            Posterior distribution of the smoothed output
+        If the extended Kalman filter is used, this yields the IEKS. In
+        any case, the result is an approximation to the maximum-a-
+        posteriori estimate.
         """
+
+        if stopcrit is None:
+            stopcrit = StoppingCriterion()
```

Comment:
there are default atol, rtol, and maxit choices.

Reply:
isnt this a problem with mutability? I agree that this would be easier, but I am a bit timid to do it like this and potentially introduce bugs that we will never find again...

What do you think?

Reply:
I think I don't really understand the potential mutability problem that you mention. What bugs could that maybe introduce?

Reply:
I am talking about https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/ which may apply here; for instance, `StoppingCriterion()` sets and reset `self.iterations` depending on whether something is accepted or not. 

Though I am not sure whether I am seeing ghosts in this context. Purely from a simplification perspective, I agree that `stopcrit=StoppingCriterion()` is simpler.

Reply:
Thanks, that makes sense! :+1: